imitation/scripts/train.py

```diff
@@ -38,130 +43,149 @@ def train_and_plot(policy_dir, env='CartPole-v1',
     tf.logging.info("exp reward: {}".format(exp_rew))
 
-  def show_plot_gen():
-    if n_gen_plot_episodes <= 0:
-      return
-
+  def ep_reward_plot_show():
+    """Render and show average episode reward plots."""
     for name in gen_ep_reward:
       plt.title(name + " Performance")
       plt.xlabel("epochs")
       plt.ylabel("Average reward per episode (n={})"
-                 .format(n_gen_plot_episodes))
+                 .format(n_episodes_per_reward_data))
       plt.plot(gen_ep_reward[name], label="avg gen ep reward", c="red")
       plt.plot(rand_ep_reward[name],
                label="avg random ep reward", c="black")
       plt.plot(exp_ep_reward[name], label="avg exp ep reward", c="blue")
       plt.legend()
       _savefig_timestamp("plot_fight_epreward_gen", interactive)
 
-  add_plot_disc(False)
-  add_plot_gen(env, "True Reward")
-  add_plot_gen(trainer.wrap_env_test_reward(env), "Learned Reward")
-
-  if n_plots_each_per_epoch <= 0:
-    n_gen_steps_per_plot = float('Inf')
-    n_disc_steps_per_plot = float('Inf')
+  if n_epochs_per_plot is not None:
+    n_plots_per_epoch = 1 / n_epochs_per_plot
   else:
-    n_gen_steps_per_plot = int(round(
-            n_gen_steps_per_epoch / n_plots_each_per_epoch))
-    n_disc_steps_per_plot = int(round(
-            n_disc_steps_per_epoch / n_plots_each_per_epoch))
+    n_plots_per_epoch = None
 
-  def train_plot_itr(steps, gen_mode, steps_per_plot):
-    nonlocal plot_idx
-    while steps > 0:
-      steps_to_train = min(steps, steps_per_plot)
-      if gen_mode:
-        trainer.train_gen(n_steps=steps_to_train)
-      else:
-        trainer.train_disc(n_steps=steps_to_train)
-      steps -= steps_to_train
-      add_plot_disc(gen_mode)
-      plot_idx += 1
-
-  for epoch_num in tqdm.trange(n_epochs, desc="epoch"):
-    train_plot_itr(n_disc_steps_per_epoch, False, n_disc_steps_per_plot)
-    train_plot_itr(n_gen_steps_per_epoch, True, n_gen_steps_per_plot)
-
-    add_plot_gen(env, "True Reward")
-    add_plot_gen(trainer.wrap_env_test_reward(env), "Learned Reward")
-
-    show_plot_disc()
-    show_plot_gen()
-    if trainer_hook_fn:
-      trainer_hook_fn(trainer)
+  def should_plot_now(epoch) -> bool:
+    """For positive epochs, returns True if a plot should be rendered now.
+
+    This also controls the frequency at which `ep_reward_plot_add_data` is
+    called, because generating those rollouts is too expensive to perform
+    every timestep.
+    """
+    assert epoch >= 1
+    if n_plots_per_epoch is None:
+      return False
+    plot_num = math.floor(n_plots_per_epoch * epoch)
+    prev_plot_num = math.floor(n_plots_per_epoch * (epoch - 1))
+    assert abs(plot_num - prev_plot_num) <= 1
+    return plot_num != prev_plot_num
+
+  # Collect data for epoch 0.
+  if n_epochs_per_plot is not None:
+    disc_plot_add_data(False)
+    ep_reward_plot_add_data(env, "True Reward")
+    ep_reward_plot_add_data(trainer.env_wrapped_test, "Learned Reward")
```

Comment:
Maybe specify in the title this is test reward? It's a bit misleading in so far as the generator is being trained with something else. Possibly log both?

Reply:
Yes, I intend on making adding better plot titles to this PR